### PR TITLE
Port to new etcd client with TLS support

### PR DESCRIPTION
**Pull request by vicki-c**
This includes:
- replacing the deprecated `go-etcd` client
- adding a `PermitPool` from the `physical` package to the etcd backend
- adding TLS support

Based on https://github.com/hashicorp/vault/pull/757

### CODE REVIEW

File: physical/etcd.go [hunk 1]
```diff
@@ -79,18 +82,54 @@ func newEtcdBackend(conf map[string]string) (Backend, error) {
 	if address, ok := conf["address"]; ok {
 		machines = address
 	}
+	machinesParsed := strings.Split(machines, EtcdMachineDelimiter)
```
Comment (jefferai) [inline physical/etcd.go]: Is there any way that we can verify that these machines are correct? e.g. if they are URLs, can you run them through `url.Parse`; if they are IP addresses can you run them through `net.ParseIP`?

**Comment by jefferai** [comment]
Looks good other than that one comment!

**Comment by jefferai** [comment]
Also, can you verify that you ran the acceptance tests successfully?

File: physical/etcd.go (hunk 1 above)
Comment (vicki-c) [inline physical/etcd.go]: yup

**Comment by jefferai** [comment]
Great! Merged!

**Comment by jcollie** [comment]
Trying out a fresh installation of 0.4.0-rc2 with a etdc 2.2.2 cluster as a backend I get the following error when I try and initialize the vault:

    Error initializing Vault: Error making API request.
    
    URL: PUT http://127.0.0.1:8200/v1/sys/init
    Code: 400. Errors:
    
    * failed to check for initialization: 100: Key not found (/vault) [7530]

**Comment by jefferai** [comment]
@vicki-c @gdb Any chance you can work with @jcollie on this? I don't know etcd very well, and this is one of only two changes to etcd since 0.3.1. It'd be helpful if you have time to dig into this with him. Otherwise I can give it a shot.

@jcollie If you are fine building from source, any chance you can git-bisect to try to track down which commit caused the problem?

**Comment by vicki-c** [comment]
yup I can look into this.

**Comment by jefferai** [comment]
@vicki-c Much appreciated. @jcollie Can you provide etcd configuration information? Are you able to bisect?

As long as we're making traction on this issue I'll hold off on 0.4 final but if it stalls for a while it'll need to get fixed post 0.4.

**Comment by vicki-c** [comment]
@jefferai pr is in.
@jcollie thanks for the report, and sorry about that!

**Comment by jefferai** [comment]
@jcollie PR #818 fixes the problem in my tests. Can you verify?

Thanks!

**Comment by jefferai** [comment]
@jcollie Note that I merged PR #818 into master, so simply verifying that master works would be sufficient, rather than pulling that branch or diff down.

**Comment by jcollie** [comment]
Took me a little while to get a go development environment set up (not really a go dev) but I was able to build master from this morning and get a vault to initialize and unseal.

**Comment by jefferai** [comment]
Great!

**Comment by maticmeznar** [comment]
Documentation at https://vaultproject.io/docs/config/index.html is missing new etcd config options. Also, "advertise_addr" seems to be mandatory and not optional, contrary to docs.

**Comment by jefferai** [comment]
@maticmeznar I'll take a look at the docs.

Regarding `advertise_addr`: The docs say "Most HA backends will attempt to determine the advertise address if not provided." The etcd backend does not support this detection, so for the etcd backend it is required.

**Pull request merged**